docs/features/patterns.md

```diff
@@ -3,104 +3,130 @@
 
 ## Is Expression
 
 The `is` operator is extended to test an expression against a *pattern*.
 
 ```antlr
-relational_expression
+is_pattern_expression
     : relational_expression 'is' pattern
     ;
+
+relational_expression
+    : is_pattern_expression
+    ;
 ```
 
 This form of *relational_expression* is in addition to the existing forms in the C# specification. It is a compile-time error if the *relational_expression* to the left of the `is` token does not designate a value or does not have a type.
 
+A *constant_pattern* appearing as the right-hand-side of an *is_pattern_expression* is syntactically restricted to be a *shift_expression*, even though a *constant_pattern* appearing elsewhere can syntactically be any *expression*. For simplicity that restriction is not shown in the grammar.
+
 Every *identifier* of the pattern introduces a new local variable that is *definitely assigned* after the `is` operator is `true` (i.e. *definitely assigned when true*).
 
+> Note: There is technically an ambiguity between *type* in an `is-expression` and *constant_pattern* in an *is_pattern_expression*, either of which might be a valid parse of a qualified identifier. We try to bind it as a type for compatibility with previous versions of the language; only if that fails do we resolve it as a constant pattern.
+
 ## Patterns
 
 Patterns are used in the `is` operator and in a *switch_statement* to express the shape of data against which incoming data is to be compared. Patterns may be recursive so that parts of the data may be matched against sub-patterns.
 
 ```antlr
 pattern
-    : type_pattern
+    : declaration_pattern
     | constant_pattern
-    | wildcard_pattern
+    | deconstruction_pattern
+    | property_pattern
+    | discard_pattern
     | var_pattern
-    | recursive_pattern
-    ;
-
-type_pattern
-    : type identifier
-    ;
-
-wildcard_pattern
-    : '*'
     ;
 
-var_pattern
-    : 'var' identifier
+declaration_pattern
+    : type simple_designation
     ;
 
 constant_pattern
-    : shift_expression
+    : constant_expression
     ;
 
-recursive_pattern
-    : positional_pattern
-    | property_pattern
-    ;
+property_pattern
+	: type? property_subpattern simple_designation?
+	;
```

Comment:
Indenting, here and in sub patterns below.

Reply:
Yes.